<!DOCTYPE html>
<html lang="en"><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
	<meta charset="utf-8">
	<title>Eric' software</title>
	<link rel="stylesheet" href="index_files/style.css" type="text/css" media="screen">
	<!-- <link rel="SHORTCUT ICON" href="./favicon.ico" /> -->
</head>
<body>
   <div id="container">

	<header><!-- header -->
		<h1><a href="index.html">Eric Frichot</a></h1>
		<h4>PhD Student, Grenoble University</h4>
 		<h4><a href="http://www-timc.imag.fr/rubrique78.html">Mathematical and computationnal biology</a>, laboratoire TIMC-IMAG</h4> 
		<img src="index_files/belledonne.jpg" alt="" height="200" width="800"> <!-- header image -->
	</header><!-- end of header -->
	
	<div id="wrapper">
	<div id="navigation">
		<div class="menu">
			<ul>
	<li><a href="index.html">Home</a></li>
	<li><a href="publi.html">Publications</a></li>
	<li><a href="soft.html">Software</a></li>
	<li><a href="conf.html">Conferences</a></li>
	<li><a href="teaching.html">Teaching</a></li>
			</ul>
		</div>
	 </div>
	
         <div id="content-wrapper">
         	<div id="content"><!-- #main content and sidebar area -->
			<!--<section id="content"--><!-- #content -->

         		<h3><strong>Softwares </strong> </h3><ul> 
			<li>
<!--<a href="http://membres-timc.imag.fr/Eric.Frichot/LEA/index.htm">LEA</a>--> 
<b>LEA</b> (coming soon)
is an R package dedicated to <b>landscape genomics</b> and <b>ecological 
association</b> tests. LEA can run analyses of population structure and genome 
scans for local adaptation. It includes statistical methods for estimating 
ancestry coefficients from large genotypic matrices and evaluating the 
number of ancestral populations (sNMF, PCA); and identifying genetic 
polymorphisms that exhibit high correlation with some environmental 
gradient or with the variables used as proxies for ecological pressures 
(LFMM), and controlling the false discovery rate. LEA is mainly based on 
optimized C programs that can scale with the dimension of very large data sets.
			</li>
			<li>
<p>
<a href="http://membres-timc.imag.fr/Eric.Frichot/snmf/index.htm">sNMF</a>: 
Inference of individual <b>admixture coefficients</b>, which is important 
for population genetic and association studies, is commonly performed using 
compute-intensive likelihood algorithms. With the availability of large 
population genomic data sets, fast versions of likelihood algorithms have 
attracted considerable attention. Reducing the computational burden of 
estimation algorithms however remains a major challenge.
<br/>
<b>sNMF</b> is a <b>fast</b> and <b>efficient</b> program for estimating 
individual admixture coefficients based on sparse non-negative matrix 
factorization and population genetics. We already successfully applied sNMF 
to large human and plant genomic data sets. The performances of sNMF 
were then compared to the likelihood algorithm implemented in the 
computer program ADMIXTURE.  Without loss of accuracy, sNMF computed 
estimates of admixture coefficients within run-times approximately 10 
to 30 times faster than those of ADMIXTURE.
        </p>
			</li>
			<li>
        <p>
	<a href="http://membres-timc.imag.fr/Eric.Frichot/lfmm/index.htm">LFMM</a>: 
        <b>Adaptation </b>to local environments often occurs through natural selection
        acting on large number of alleles, each having a weak phenotypic effect.
        One way to detect those alleles is by identifying genetic polymorphisms
        that exhibit high <b>correlation</b> with some <b>environmental gradient</b> or with
        the variables used as proxies for <b>ecological</b> pressures.
	<br/>
        The <b>LFMM computer program</b> includes an integrated framework based on
        population genetics,
        ecological modeling and statistical techniques for screening genomes for
        signatures of <b>local adaptation</b>.
	<br/>
        It implements <b>fast algorithms</b> using latent
        factor mixed models based on a variant of Bayesian principal component
        analysis in which residual <b>population structure</b> is introduced via unobserved
        factors.
        </p>
			</li>
			</ul><table style="width: 600px; height: 200px;" border="0">
		        <tbody>
                	<tr>

	                

				

	        		

				


	                </tr>
		        </tbody>
		        </table>
		</div>
	</div>

<div id="footer">
<a href="http://www-timc.imag.fr/"><img style="width: 97px; height: 72px; left" src="index_files/logotimc.gif" alt="logotimc"></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="http://www.rhonealpes.fr/"><img alt="ANR" src="lfmm/images/region.jpg">
</a></div>

</a></div></div></body></html>
